torch/distributed/_composable/fsdp/_fsdp_collectives.py

```diff
@@ -105,6 +110,116 @@ def foreach_all_gather_copy_out(
         torch._foreach_copy_(out, splits)  # one `copy_` per parameter
 
 
+@torch.no_grad()
+def foreach_reduce_scatter(
+    fsdp_params: List[FSDPParam],
+    unsharded_grads: List[torch.Tensor],
+    group: dist.ProcessGroup,
+    reduce_scatter_stream: torch.cuda.Stream,
+    orig_dtype: torch.dtype,
+    reduce_dtype: Optional[torch.dtype],
+    device: torch.device,
+    predivide_factor: float,
+    postdivide_factor: float,
+) -> torch.cuda.Event:
+    """
+    ``unsharded_grads`` owns the references to the gradients computed by
+    autograd, so clearing the list frees the gradients.
+    """
+    grad_dtypes = {grad.dtype for grad in unsharded_grads}
+    if len(grad_dtypes) != 1:
+        # Check this at runtime since it could be a real runtime error if e.g.
+        # fp8 weights do not produce the correct higher precision gradients
+        _raise_assert_with_print(
+            f"FSDP reduce-scatter expects uniform gradient dtype but got {grad_dtypes}"
+        )
+    grad_dtype = unsharded_grads[0].dtype
+    reduce_dtype = reduce_dtype or grad_dtype
+    world_size = group.size()
+    padded_unsharded_sizes = tuple(
+        _get_dim0_padded_size(grad.size(), world_size) for grad in unsharded_grads
+    )
+    reduce_scatter_input_numel = sum(s.numel() for s in padded_unsharded_sizes)
+    reduce_scatter_output_numel = reduce_scatter_input_numel // world_size
+    current_stream = torch.cuda.current_stream()
+    reduce_scatter_stream.wait_stream(current_stream)
+    with torch.cuda.stream(reduce_scatter_stream):
+        reduce_scatter_input = torch.empty(
+            (reduce_scatter_input_numel,), dtype=reduce_dtype, device=device
+        )
+        foreach_reduce_scatter_copy_in(
+            fsdp_params, unsharded_grads, reduce_scatter_input, world_size
+        )
+        _div_if_needed(reduce_scatter_input, predivide_factor)
```

Comment:
Depending on whether we end up generating triton kernels for reduce-scatter-copy-in, we might want to move this to a different place.

In case we won't be generating a triton kernel for reduce-scatter-copy-in (e.g. due to triton's limitation in handling the workload), we might want to perform the division prior to the copy-in. This way it might be picked up as an epilogue for gradient function (we can't fuse two kernels if there's an extern kernel in-between). Or we could fuse the division with reduce-scatter-copy-in.

If we can generate an efficient triton kernel for reduce-scatter-copy-in, the order won't matter.

This is not relevant to the PR. But it would be nice to note down so we know where we can hunt for perf later.

Reply:
(This is probably pre-mature optimization thinking so don't worry too much about it)